modules/dlm/src/main/java/org/elasticsearch/dlm/DataLifecycleService.java

```diff
@@ -250,6 +340,41 @@ private void maybeExecuteRetention(ClusterState state, DataStream dataStream) {
                 );
             }
         }
+        return indicesToBeRemoved;
+    }
+
+    /*
+     * This method force merges any indices in the datastream that are not in the indicesToExclude set and that have not be force merged
+     * previously. Before force merging, it writes custom metadata as a marker in the cluster state so that we know a force merge is in
+     * process. And it writes another marker in the cluster state upon completion of the force merge.
+     */
+    private void maybeExecuteForceMerge(ClusterState state, DataStream dataStream, Set<Index> indicesToExclude) {
+        List<Index> readOnlyIndices = dataStream.getIndices().stream().filter(index -> indicesToExclude.contains(index) == false).toList();
+        Metadata metadata = state.metadata();
+        for (Index index : readOnlyIndices) {
+            if (dataStream.isIndexManagedByDLM(index, state.metadata()::index)) {
+                IndexMetadata backingIndex = metadata.index(index);
+                assert backingIndex != null : "the data stream backing indices must exist";
+                String indexName = index.getName();
+                boolean alreadyForceMerged = isForceMergeComplete(backingIndex);
+                if (alreadyForceMerged) {
+                    logger.trace("Already force merged {}", indexName);
```

Comment:
Should we `continue` here? (no need to check failures and such anymore for this index)

Reply:
Oh good point. I'd added those ifs for logging after I'd written the rest of it, and they make the later if obsolete.